prefect_kubernetes/jobs.py

```diff
@@ -299,3 +309,205 @@ def kubernetes_orchestrator():
+        ) as core_v1_client:
+            completed = False
+
+            elapsed_time = 0
+
+            while not completed:
+                job_expired = (
+                    elapsed_time > self._kubernetes_job.timeout_seconds
+                    if self._kubernetes_job.timeout_seconds
+                    else False
+                )
+                if job_expired:
+                    raise TimeoutError(f"Job timed out after {elapsed_time} seconds.")
+
+                latest_v1_job = await run_sync_in_worker_thread(
+                    batch_v1_client.read_namespaced_job_status,
+                    name=self._kubernetes_job.v1_job.metadata.name,
+                    namespace=self._kubernetes_job.namespace,
+                    **self._kubernetes_job.api_kwargs,
+                )
+
+                v1_pod_list = await run_sync_in_worker_thread(
+                    core_v1_client.list_namespaced_pod,
+                    namespace=self._kubernetes_job.namespace,
+                    label_selector=(
+                        f"controller-uid="
+                        f"{latest_v1_job.metadata.labels['controller-uid']}"
+                    ),
+                    **self._kubernetes_job.api_kwargs,
+                )
+                for pod in v1_pod_list.items:
+                    pod_name = pod.metadata.name
+
+                    if pod.status.phase == "Pending" or pod_name in self.pod_logs:
+                        continue
+
+                    self.logger.info(f"Capturing logs for pod {pod_name!r}.")
+
+                    self.pod_logs[pod_name] = await run_sync_in_worker_thread(
+                        core_v1_client.read_namespaced_pod_log,
+                        namespace=self._kubernetes_job.namespace,
+                        name=pod_name,
+                        container=latest_v1_job.spec.template.spec.containers[0].name,
+                        **self._kubernetes_job.api_kwargs,
+                    )
+
+                if latest_v1_job.status.active:
+                    await sleep(self._kubernetes_job.interval_seconds)
+                    if self._kubernetes_job.timeout_seconds:
+                        elapsed_time += self._kubernetes_job.interval_seconds
+                elif latest_v1_job.status.failed:
+                    raise RuntimeError(
+                        f"Job {latest_v1_job.metadata.name!r} failed, check the "
+                        "Kubernetes pod logs for more information."
+                    )
+                elif latest_v1_job.status.succeeded:
+                    completed = True
+                    self.logger.info(
+                        f"Job {latest_v1_job.metadata.name!r} has completed."
+                    )
+
+    async def fetch_result(self) -> Dict[str, Any]:
+        """Fetch the results of the job.
+
+        Returns:
+            The logs from each of the pods in the job.
+        """
+        if self._kubernetes_job.delete_after_completion:
```

Comment:
Thinking from the user perspective; what if I forget to call `wait_for_completion`? This would delete the job and return empty logs, which can be misleading. Should we raise an error if self.pod_logs is empty?

Reply:
i don't think we should raise on empty pod_logs, because pods could literally just not emit logs as a part of normal, user-specified operation

Reply:
What if we set `self.pod_logs = None`, and when `wait_for_completion` is called, it sets to `{}` if it's None? Then raise here if it's None?

Reply:
good idea!